pkg/virt-operator/resource/apply/certificates.go

```diff
@@ -1,31 +1,96 @@
 package apply
 
 import (
+	"time"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	k8sv1 "kubevirt.io/client-go/api/v1"
 )
 
 func getCADuration(config *k8sv1.KubeVirtSelfSignConfiguration) *metav1.Duration {
-	if config == nil || config.CARotateInterval == nil {
-		return &metav1.Duration{Duration: Duration7d}
+	defaultDuration := &metav1.Duration{Duration: Duration7d}
+
+	if config == nil {
+		return defaultDuration
+	}
+
+	// deprecated, but takes priority to provide a smooth upgrade path
+	if config.CARotateInterval != nil {
+		return config.CARotateInterval
+	}
+	if config.CA != nil && config.CA.Duration != nil {
+		return config.CA.Duration
+	}
+
+	return defaultDuration
+}
+
+func getCARenewalPeriod(config *k8sv1.KubeVirtSelfSignConfiguration) *metav1.Duration {
+	caDuration := getCADuration(config)
+	defaultDuration := &metav1.Duration{Duration: time.Duration(float64(caDuration.Duration) * 0.8)}
+
+	if config == nil {
+		return defaultDuration
 	}
 
-	return config.CARotateInterval
+	if config.CA != nil && config.CA.RenewBefore != nil {
+		return config.CA.RenewBefore
+	}
+
+	return defaultDuration
 }
 
 func getCAOverlapTime(config *k8sv1.KubeVirtSelfSignConfiguration) *metav1.Duration {
-	if config == nil || config.CAOverlapInterval == nil {
-		return &metav1.Duration{Duration: Duration1d}
+	defaultDuration := &metav1.Duration{Duration: Duration1d}
+
+	if config == nil {
+		return defaultDuration
+	}
+
+	// deprecated, but takes priority to provide a smooth upgrade path
+	if config.CAOverlapInterval != nil {
+		return config.CAOverlapInterval
+	}
+
+	if config.CA == nil || config.CA.Duration == nil || config.CA.RenewBefore == nil {
+		return defaultDuration
 	}
 
-	return config.CAOverlapInterval
+	caDuration := getCADuration(config)
+	caRenewalPeriod := getCARenewalPeriod(config)
+	return &metav1.Duration{Duration: caDuration.Duration - caRenewalPeriod.Duration}
```

Comment:
@stu-gott Is there any place that makes sure that `caDuration` will always be higher than `caRenewalPeriod`?

Reply:
Yes. This exact same code is used to generate both the CA and the cert.